include/micm/solver/cuda_lu_decomposition.cuh

```diff
@@ -3,13 +3,27 @@
 // SPDX-License-Identifier: Apache-2.0
 #pragma once
 #include <micm/util/cuda_param.hpp>
-#include <vector>
 namespace micm
 {
   namespace cuda
   {
+    /// This is the host function that will call the CUDA kernel
+    ///   to perform LU decomposition on the device
+    /// Note that referencing "devstruct" as "LuDecomposeConst&"
+    ///   will cause an error like `error: qualifiers dropped in 
+    ///   binding reference of type "LuDecomposeConst &" to 
+    ///   initializer of type "const LuDecomposeConst"`
     std::chrono::nanoseconds DecomposeKernelDriver(
             CudaSparseMatrixParam& sparseMatrix, 
-            CudaSolverParam& solver);
-  }  // namespace cuda
-}  // namespace micm
+            LuDecomposeConst devstruct);
```

Comment:
@mattldawson @K20shores I got a compilation error when using `LuDecomposeConst&` and the error message is listed in the comment. I switch to passing by value to avoid this error but let me know if there is a better way to do so. I also do not understand the error message here.

Reply:
I don't see the error message - is it attached to another comment?

Reply:
Haha, it is not clear in the comment but here it is: `error: qualifiers dropped in binding reference of type "LuDecomposeConst &" to initializer of type "const LuDecomposeConst"`

Reply:
ah, ok - I think you can change the function argument to be `const LuDecomposeConst&` and it should work (as long as you aren't modifying it in the function, which I don't think is the case)

Reply:
so in this case, do I have to declare `const LuDecomposeConst devstruct_` in the `CudaLuDecomposition` class, too?

Reply:
no, if the function takes a `const` it will also work a non-`const` argument - it's essentially just a promise that the function won't try to modify the referenced argument, so it's safe to pass things that can't be modified

Reply:
Oh I see. Thanks for the clarification. I thought C++ must match the variable type exactly but  the keyword `const` seems to be an exception?

Reply:
I think the difference is that `const` is a qualifier as opposed to being part of the type itself

Reply:
I add the `const` qualifier to the `LuDecomposeConst` type and it works for the `DecomposeKernelDriver` function. However, if I do the same thing for the CUDA kernel `LuDecomposeConst`, the compilation passes but the gpu test of `cuda_lu_decomposition` will hang. Thus I still use passing by value for that kernel.

Is passing by reference a new feature in C++ standard? I see a compilation warning saying `nvcc warning : The -std=c++20 flag is not supported with the configured host compiler. Flag will be ignored.` and maybe this is the reason why CUDA kernel does not work with reference?

Reply:
@sjsprecious pass by value and pass by reference is not new in c++. To my knowledge it has been in the language from the beginning. The flag you are getting seems to suggest that yet again the configuration of the nividia compiler on the HPC machines is incorrect.

Reply:
According to this stackflow https://stackoverflow.com/questions/17794492/reference-argument-on-cuda), we need a pointer/copy to the device memory for the CUDA kernel argument. Thus the reference does not work for CUDA kernel in this case.